nexus/assets/js/home.js

```diff
@@ -0,0 +1,144 @@
+let channel = undefined;
+let pc = undefined;
+let localTracksAdded = false;
+let candidates = [];
+
+async function createPeerConnection() {
+  pc = new RTCPeerConnection(pcConfig);
+
+  pc.ontrack = event => {
+    if (event.track.kind == "video") {
+      console.log("Creating new video element");
+
+      const trackId = event.track.id;
+      const videoPlayer = document.createElement("video");
+      videoPlayer.srcObject = event.streams[0];
+      videoPlayer.autoplay = true;
+      videoPlayer.classList.add("m-auto", "rounded-xl", "max-h-full", "max-w-full");
+
+      videoPlayerWrapper.appendChild(videoPlayer);
+
+      event.track.onended = _ => {
+        console.log("Track ended: " + trackId);
+        videoPlayerWrapper.removeChild(videoPlayer);
+      }
+    } else {
+        // Audio tracks are associated with the stream (`event.streams[0]`) and require no separate actions
+        console.log("New audio track added");
+    }
+  }
+
+  pc.onicegatheringstatechange = () => console.log("Gathering state change: " + pc.iceGatheringState);
+  pc.onconnectionstatechange = () => console.log("Connection state change: " + pc.connectionState);
+  pc.onicecandidate = event => {
+    if (event.candidate == null) {
+      console.log("Gathering candidates complete");
+      forwardSdpAnswer();
+      return;
+    }
+
+    const candidate = JSON.stringify(event.candidate);
+    console.log("Sending ICE candidate: " + candidate);
+    channel.push("ice_candidate", {body: candidate});
+  }
+}
+
+async function setupLocalMedia() {
+  console.log("Setting up local media stream");
+  // ask for permissions
+  localStream = await navigator.mediaDevices.getUserMedia({ video: true, audio: true });
+  console.log(`Obtained stream with id: ${localStream.id}`);
+
+  setupPreview();
+}
+
+function setupPreview() {
+  // flip the preview horizontally
+  ownVideoPlayer.style.cssText = "-moz-transform: scale(-1, 1); \
+-webkit-transform: scale(-1, 1); -o-transform: scale(-1, 1); \
+transform: scale(-1, 1); filter: FlipH;";
+
+  ownVideoPlayer.srcObject = localStream;
+}
+
+async function joinChannel(streamPromise) {
+  const socket = new Socket("/socket");
+  socket.connect();
+  channel = socket.channel(`peer:signalling`);
+
+  channel.onError(() => {
+    socket.disconnect();
+    window.location.reload();
+  });
+  channel.onClose(() => {
+    socket.disconnect();
+    window.location.reload();
+  });
+
+  channel.on("sdp_offer", payload => {
+    const sdpOffer = payload.body;
+    
+    console.log("SDP offer received");
+    streamPromise.then(async (_) => {
+      await applySdpOffer(sdpOffer);
+      if (pc.connectionState == "connected") {
+        forwardSdpAnswer();
+      }
+    });
+  });
+
+  channel.on("ice_candidate", payload => {
+    const candidate = JSON.parse(payload.body);
+    console.log("Received ICE candidate: " + payload.body);
+    if (pc.remoteDescription !== null) {
+      pc.addIceCandidate(candidate);
+    } else {
```

Comment:
This should never happen as we send SDP offer immediately after generating it on the BE :thinking: